### PR TITLE
Return more specific errors about invalid descriptors

### PR DESCRIPTION
**Pull request by achow101**
Because it's nigh impossible to figure out what is wrong with a descriptor otherwise.

### CODE REVIEW

**Comment by sipa** [comment]
Descriptors are hard to parse, they should be hard to write also.

Eh, I mean, big concept ACK.

**Comment by jb55** [comment]
Concept ACK

**Comment by Sjors** [comment]
Concept ACK

**Review by promag — COMMENTED** [review]
Concept ACK, nice to have a detailed error.

**Comment by hugohn** [comment]
Concept ACK.

File: src/script/descriptor.cpp
Comment (Sjors) [inline src/script/descriptor.cpp]: Nit: "not valid" instead of "out of range"? Or split the check?

File: test/functional/wallet_importmulti.py
Comment (Sjors) [inline test/functional/wallet_importmulti.py]: nit: make descriptor errors start with lower case?

**Review by Sjors — APPROVED** [review]
ACK dae2590

This introduces a fun little trick if you're too lazy to call `getdescriptorinfo` to obtain the checkum: you can instead set the checksum to `00000000`.
```
$ bitcoin-cli importmulti '[{"desc": "sh(wpkh(03fff97bd5755eeea420453a14355235d382f6472f8568a18b2f057a1460297556))#00000000", "timestamp": "now"}]'
[
  {
    "success": false,
    "error": {
      "code": -5,
      "message": "Descriptor is invalid, Provided checksum '00000000' does not match computed checksum 'qkrrc7je'"
    }
  }
]

```

File: test/functional/wallet_importmulti.py
Comment (achow101) [reply test/functional/wallet_importmulti.py]: Nah.

File: src/script/descriptor.cpp
Comment (achow101) [reply src/script/descriptor.cpp]: I've split the check

File: src/test/descriptor_tests.cpp
Comment (practicalswift) [inline src/test/descriptor_tests.cpp]: Nit: `expected_error` could be const ref?

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: src/script/descriptor.cpp
Comment (promag) [inline src/script/descriptor.cpp]: Missing test.

File: test/functional/wallet_importmulti.py
Comment (promag) [reply test/functional/wallet_importmulti.py]: Agree with @Sjors, is there a reason to not do so? I was going to suggest to drop "Descriptor is invalid, " prefix.

**Review by promag — COMMENTED** [review]
I think I've bookmarked all errors without a test - not saying you should add them here, I'm happy to improve in a follow up.

File: src/script/descriptor.cpp
Comment (achow101) [reply src/script/descriptor.cpp]: I don't think this check is actually possible to hit due to how `Expr` parses things earlier.

**Comment by achow101** [comment]
I've added more tests and also cleaned up a few typos in error messages.

File: src/test/descriptor_tests.cpp
Comment (achow101) [reply src/test/descriptor_tests.cpp]: Done

File: test/functional/wallet_importmulti.py
Comment (achow101) [reply test/functional/wallet_importmulti.py]: I've just removed the prefixes.

**Comment by Sjors** [comment]
reACK a1b47955838f1151fc1904d0ff3bc7751b9fa7ca

**Comment by jb55** [comment]
Code Review ACK a1b47955838f1151fc1904d0ff3bc7751b9fa7ca

File: src/script/descriptor.cpp
Comment (instagibbs) [inline src/script/descriptor.cpp]: where would this case be hit previously?

File: src/script/descriptor.cpp
Comment (instagibbs) [inline src/script/descriptor.cpp]: where would this case be hit previously?

File: src/script/descriptor.cpp
Comment (instagibbs) [inline src/script/descriptor.cpp]: nit: I'd prefer it also checks for `!Func` explicitly, since there are a number of combinations, and a generic error is preferable to a specific error that's an implementation mistake.

**Review by instagibbs — APPROVED** [review]
tACK

the errors I encountered during initial usage of descriptors are all caught here, and enough information to quickly figure out what I did wrong.

f.e.

```
getdescriptorinfo "multi(1, 03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd,04a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd5b8dec5235a0fa8722476c7709c02559e3aa73aa03918ba2d492eea75abea235)"
error code: -5
error message:
key ' 03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd' is not valid
```
^ space after comma

File: src/script/descriptor.cpp
Comment (achow101) [reply src/script/descriptor.cpp]: It would fail below at the extkey decoding stuff.

File: src/script/descriptor.cpp
Comment (achow101) [reply src/script/descriptor.cpp]: It would just get to the end of the function and hit `return nullptr`.

**Comment by Sjors** [comment]
If you need to make any changes, I suggest rebasing due to #14934 to be on the safe side. It still compiles and tests pass for me when I do that rebase locally though.

File: src/script/descriptor.cpp
Comment (achow101) [reply src/script/descriptor.cpp]: I don't think it can check that there is no function.

**Comment by meshcollider** [comment]
Sorry about the rebase again due to #15986 

Once this is rebased I'll review and merge

**Comment by achow101** [comment]
Rebased

File: src/script/descriptor.cpp
Comment (Sjors) [reply src/script/descriptor.cpp]: Now it's no longer split.

File: test/functional/wallet_importmulti.py
Comment (Sjors) [reply test/functional/wallet_importmulti.py]: They're back

**Review by Sjors — CHANGES_REQUESTED** [review]
Only the first commit changed in this rebase, but you lost a few of your recent fixes in the process.

**Comment by achow101** [comment]
> Only the first commit changed in this rebase, but you lost a few of your recent fixes in the process.

No??? Are you sure you're looking at the right commits? The latest diff includes everything that was commented on earlier.

File: test/functional/wallet_importmulti.py
Comment (achow101) [reply test/functional/wallet_importmulti.py]: > They're back

No??

File: src/script/descriptor.cpp
Comment (achow101) [reply src/script/descriptor.cpp]: > Now it's no longer split.

No??

**Comment by Sjors** [comment]
Ah wait, I see these changes happen in 625534d7b1417da926f1ced600855ea818d6e01e.

ACK 787c9ec

File: src/script/descriptor.cpp
Comment (meshcollider) [inline src/script/descriptor.cpp]: mismatched `'` (also in the corresponding test case)

**Review by meshcollider — COMMENTED** [review]
Code review ACK 787c9ec0c383cefb83a95433311a27f9c220dca0

This is a nice change, one tiny nit but it is pretty insignificant so I'll merge this tomorrow if you don't have time to fix it now

**Pull request merged**
